CSharp/JobPrep/Program.cs

```diff
@@ -0,0 +1,333 @@
+
+                    // launch N threads to get job prep task status of the nodes in the queue
+                    for (int i = 0; i < parallelism; i++)
+                        tasks.Add(Task.Run(() =>
+                        {
+                            while (!ct.IsCancellationRequested)
+                            {
+                                string vm;
+                                try {
+                                    vm = vms.Take(ct);
+                                } catch // if token cancelled, abort the thread
+                                {
+                                    break;
+                                }
+                                Log("start polling vm", vm);
+                                try
+                                {
+                                    var state = pm.GetJobPreparationTaskStatus(pool, vm, workitem, job);
+                                    Log("got vm state", vm);
+                                    if (state.State == JobPreparationTaskState.Completed) // job prep task completed on the thread
+                                    {
+                                        if (state.ExitCode == 0)
+                                            Interlocked.Increment(ref nb_succeeded);
+                                        else
+                                            Interlocked.Increment(ref nb_failed);
+                                        Log("==== job prep done ====", vm);
+                                        if (stop(nb_succeeded, nb_failed))
+                                            cts.Cancel();
+                                    }
+                                    else // job prep still running
+                                    {
+                                        Log("state not completed", vm);
+                                        // Sleep for some time and put the VM back to the queue
+                                        Task.Run(() => { Task.Delay(freq); if (!ct.IsCancellationRequested) { Log("add vm back", vm); vms.Add(vm); } });
+                                    }
+                                }
+                                catch // job prep task status not found - not run on this VM
+                                {
+                                    Log("error query", vm);
+                                    Task.Run(() => {Task.Delay(freq); if (!ct.IsCancellationRequested) { Log("add vm back", vm); vms.Add(vm); } });
+                                }
+                            }
+                            Log("cancel");
+                        }));
+                    Task.WaitAll(tasks.ToArray(), timeout);
+                    Log("all done");
+                    return succeed(nb_succeeded, nb_failed);
+                }
+            }
+        }
+
+        private static void AddWork(IBatchClient client)
+        {
+            using (IPoolManager pm = client.OpenPoolManager()) {
```

Comment:
You can put both of these usings together like so:
using(IPoolManager...)
using(IWorkItemManager...)
{

}